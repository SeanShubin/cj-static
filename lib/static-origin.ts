import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as cloudfront from '@aws-cdk/aws-cloudfront';
import { ManagedPolicy, PolicyStatement, Effect } from '@aws-cdk/aws-iam';
export interface IStaticOrigin {

    readonly store: s3.Bucket,
    readonly accessId: cloudfront.OriginAccessIdentity,
    readonly writeAccessPolicy: ManagedPolicy,

}


export class StaticOrigin extends cdk.Construct implements IStaticOrigin {
    writeAccessPolicy: ManagedPolicy;
    store: s3.Bucket;
    accessId: cloudfront.OriginAccessIdentity;

    constructor(scope: cdk.Construct) {
        super(scope, "Origin");

        this.store = new s3.Bucket(this, 'Content');

        this.accessId = new cloudfront.OriginAccessIdentity(this, 'access-id');

        this.writeAccessPolicy = new WriteAccessPolicy(this, this.store);
        new AccessIdReadPolicy(this, this.store, this.accessId);

    }
}


class AccessIdBucketReadAnyStatement extends PolicyStatement {
    constructor(bucket: s3.Bucket, accessId: cloudfront.OriginAccessIdentity) {
        super();
        this.effect = Effect.ALLOW;
        this.addActions('s3:GetObject');
        this.addResources(bucket.arnForObjects('*'));
        this.addCanonicalUserPrincipal(accessId.cloudFrontOriginAccessIdentityS3CanonicalUserId);
    }
}

class BucketWriteAnyStatement extends PolicyStatement {
    constructor(bucket: s3.Bucket) {
        super();
        this.effect = Effect.ALLOW;
        this.addActions(
            's3:PutObject',
            's3:DeleteObject',
            's3:GetObject',
            's3:ListBucket',
            's3:AbortMultipartUpload',
            's3:ListBucketMultipartUploads',
            's3:ListMultipartUploadParts');
        this.addResources(bucket.arnForObjects('*'), bucket.bucketArn);
    }
}


export class AccessIdReadPolicy extends s3.BucketPolicy {
    bucket: s3.Bucket;
    accessId: cloudfront.OriginAccessIdentity;

    constructor(scope: cdk.Construct, bucket: s3.Bucket, accessId: cloudfront.OriginAccessIdentity) {
        super(scope, bucket.node.id + "AccessIdReadAccess", {
            bucket: bucket
        });
        this.document.addStatements(new AccessIdBucketReadAnyStatement(bucket, accessId));
        this.bucket = bucket;
        this.accessId = accessId;
    }
}

export class WriteAccessPolicy extends ManagedPolicy {

    bucket: s3.Bucket;

    constructor(scope: cdk.Construct, bucket: s3.Bucket) {
        super(scope, bucket.node.id + "WriteAccess", {
            statements: [new BucketWriteAnyStatement(bucket)]
        });
        this.bucket = bucket;
    }
}


