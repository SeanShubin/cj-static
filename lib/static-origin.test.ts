import {StaticOrigin, WriteAccessPolicy, AccessIdReadPolicy} from './static-origin';
import { expect as expectCDK, haveResource, SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import * as cloudfront from '@aws-cdk/aws-cloudfront';

test('Has Content Bucket', () => {

    const stack = new Stack();

    new StaticOrigin(stack);

    expectCDK(stack).to(haveResource('AWS::S3::Bucket'));
});

test('Has Origin Access Id', () => {

    const stack = new Stack();

    new StaticOrigin(stack);

    expectCDK(stack).to(haveResource('AWS::CloudFront::CloudFrontOriginAccessIdentity'));
});

test('Has Origin Access Read Bucket Policy', () => {

    // when
    const origin = new StaticOrigin(new Stack());

    // then
    // a. we find a single AcccessIdReadPolicy
    const matching = origin.node.children.filter(c => c instanceof AccessIdReadPolicy);
    expect(matching.length).toBe(1);

    // b. policy is attached to content bucket
    expect((matching[0] as AccessIdReadPolicy).bucket).toBe(origin.store);

    // c. policy is attched to origin access id
    expect((matching[0] as AccessIdReadPolicy).accessId).toBe(origin.accessId);

});

test('Has Content Write Access Policy', () => {

    // when
    const origin = new StaticOrigin(new Stack());

    // then
    // a. we find a single WriteAccessPolicy in the construct
    const matching = origin.node.children.filter(c => c instanceof WriteAccessPolicy);
    expect(matching.length).toBe(1);

    // b. policy bucket is the origin content bucket
    expect((matching[0] as WriteAccessPolicy).bucket).toBe(origin.store);

});



/* -------------------------------------------------------------------------------------------------------------
 * Experimental snapshot testing
 *
 * Using snapshots to prevent regressions on highly technical, high-risk items such as security policy. These
 * are the least likely to benefit from TDD since they are data, not behavior. However, they will necessarily be
 * driven by wizard ops work which is kind of lousy.
 */


test ('Snapshot: Write Access Policy', () => {

    const stack = new Stack();
    const bucket = new Bucket(stack, "test");

    const policy = new WriteAccessPolicy(stack, bucket);

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

});

test ('Snapshot: Access Id Read Access Policy', () => {

    const stack = new Stack();
    const bucket = new Bucket(stack, "test");
    const id = new cloudfront.OriginAccessIdentity(stack, "testid");

    const policy = new AccessIdReadPolicy(stack, bucket, id);

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

});
