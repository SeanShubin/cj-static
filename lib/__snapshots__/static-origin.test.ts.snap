// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: Access Id Read Access Policy 1`] = `
Object {
  "Resources": Object {
    "testAF53AC38": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testAccessIdReadAccess07CF1021": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "testAF53AC38",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "testidCD9C1A02",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testAF53AC38",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testidCD9C1A02": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
}
`;

exports[`Snapshot: Write Access Policy 1`] = `
Object {
  "Resources": Object {
    "testAF53AC38": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testWriteAccess01E2B6E6": Object {
      "Properties": Object {
        "Description": "",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:PutObject",
                "s3:DeleteObject",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:AbortMultipartUpload",
                "s3:ListBucketMultipartUploads",
                "s3:ListMultipartUploadParts",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testAF53AC38",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
                Object {
                  "Fn::GetAtt": Array [
                    "testAF53AC38",
                    "Arn",
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
  },
}
`;
